src/collection/historical_signatures/signature_collector.py

```diff
@@ -0,0 +1,215 @@
+                    slot=signature.slot,
+                    block_time=signature.block_time,
+                    collection_stream=db.CollectionStreamTypes.SIGNATURE
+                )
+                session.add(tx_status_record)
+                session.flush()  # Flush here to get the ID
+                # store errors and/or memos to db if any
+                if signature.err:
+                    error = signature.err.to_json() if not isinstance(signature.err, TransactionErrorFieldless) else ""
+                    tx_error_record = db.TransactionStatusError(
+                        error_body=error,
+                        tx_signatures_id=tx_status_record.id,
+                    )
+                    session.add(tx_error_record)
+                if signature.memo:
+                    tx_memo_record = db.TransactionStatusMemo(
+                        memo_body=signature.memo,
+                        tx_signatures_id=tx_status_record.id,
+                    )
+                    session.add(tx_memo_record)
+
+                self._oldest_signature = signature.signature
+            session.commit()
+        LOGGER.info("Signatures successfully stored to database.")
+
+    def _fetch_signatures(self) -> None:
+        """
+        Fetch transaction signatures.
+        Fetch only signatures that occurs before `self._oldest_signature` for the given protocol.
+        """
+        self._rate_limit_calls()
+
+        try:
+            response = self.solana_client.get_signatures_for_address(
+                solana.rpc.api.Pubkey.from_string(self.protocol),
+                limit=TX_BATCH_SIZE,
+                before=self._oldest_signature,
+            )
+            signatures = response.value
+        except SolanaRpcException as e:
+            LOGGER.error(f"SolanaRpcException: {e}")
+            time.sleep(0.5)
+            self._fetch_signatures()
+            return
+
+        if len(signatures) < TX_BATCH_SIZE:
+            self._signatures_completed = True
+
+        # Get the oldest block for which it is certain that we fetched all signatures.
+        unique_slots = sorted(list({i.slot for i in signatures}))
+        if len(unique_slots) < 2:  # TODO: take care of blocks hat contain more than 1000 transactions for one protocol  # pylint: disable=W0511
+            LOGGER.warning(
+                "Last batch for protocol = {} contains only signatures from a single slot = {}.".format(
+                    self.protocol,
+                    unique_slots[0],
+                )
+            )
+            self._oldest_completed_slot = unique_slots[0]
+        else:
+            second_oldest_slot = unique_slots[1]
+            self._oldest_completed_slot = second_oldest_slot
+
+        self._collected_signatures = signatures
+
+    def _get_watershed_block_signature(self) -> None:
+        """
+        Get signature from watershed block. Signature does not have to be from relevant protocol.
```

Comment:
If the signature does not have to be from relevant protocol, then why do we filter based on the protocol in the query?

Reply:
good catch. fixed.